Move Module Configuration doc to Component Storage doc page (#977)
PrestaShop/docs: src/content/1.7/development/components/configuration.md

---
title: Configuration storage
weight: 80
aliases:
  - /1.7/modules/concepts/configuration/
---

# Configuration storage

The Configuration storage of PrestaShop aims to address two needs.

First, there are multiple situations where a single value has to be stored and be available easily. It can be a boolean to control order mode, it can be a limit for timeout: a simple key => value(s) storage is needed to manage multiple settings in PrestaShop.

Secondly, modules might sometimes need a simple storage system, and they should not have to create a new SQL table for such a simple need.

To address these needs, the Configuration storage of PrestaShop uses a SQL table `ps_configuration` and allows simple fetch, update and deletion of values in it.

> {{< figure src="../img/configuration_sql_structure.png" title="Configuration SQL structure" >}}

Reviewer: Don't you think it would be better to store this within the components section? (If yes, you will need to move the file)

```suggestion
{{< figure src="img/configuration_sql_structure.png" title="Configuration SQL structure" >}}
```

Author: I'm not to understand, the PNG image is already inside `components` section.
```
src/content/1.7/development/components/img/configuration_sql_structure.png 
```

To be able to use `img/configuration_sql_structure.png` filepath we would need to create a `configuration` subdirectory inside `component` directory, and its only purpose would be to contain `img`.

is it worth it?

Reviewer: Oh right, forgot it wasn't a directory.

Each value stored in the Configuration component can be controlled through a given key, is compatible with Multishop (as it can be overridden for a shop or a group of shops) and has timestamps.

> The storage is performed in `Configuration` class.
> 
> We recommend to access it through service `prestashop.adapter.legacy.configuration` (class `PrestaShop\PrestaShop\Adapter`), this is the interface allowing each developer to get or store configuration data.

Reviewer: ```suggestion
The storage is managed by the `Configuration` class.

{{% notice tip %}}
Although you can access `Configuration` statically from any class, we recommend injecting the `prestashop.adapter.legacy.configuration` service instead.
{{% /notice %}}
```

## Usage

### Store configuration data

```php
Configuration::updateValue(string $key, mixed $value);
```

`$key` will be the reference to your data to reuse later.
`$value` can be any scalar type (int, float, string, bool) or an array.

{{% notice tip %}}

> 
> **Multi language**.

Reviewer: Remove extra line break
```suggestion
**Multi language**.
```

---
title: Configuration storage

> weight: 80

Reviewer: ```suggestion
weight: 80
aliases:
  - /1.7/modules/concepts/configuration/
```

---

# Configuration storage

The Configuration storage of PrestaShop aims to address two needs.

First, there are multiple situations where a single value has to be stored and be available easily. It can be a boolean to control order mode, it can be a limit for timeout: a simple key => value(s) storage is needed to manage multiple settings in PrestaShop.

Secondly, modules might sometimes need a simple storage system, and they should not have to create a new SQL table for such a simple need.

To address these needs, the Configuration storage of PrestaShop uses a SQL table `ps_configuration` and allows simple fetch, update and deletion of values in it.

> {{< figure src="/1.7/development/components/img/configuration_sql_structure.png" title="Configuration SQL structure" >}}

Reviewer: You can use relative paths
```suggestion
{{< figure src="img/configuration_sql_structure.png" title="Configuration SQL structure" >}}
```

---
title: Configuration storage
weight: 80
---

# Configuration storage

The Configuration storage of PrestaShop aims to address two needs.

> First, there is multiple situations where a single value has to be stored and be available easily. It can be a boolean to control order mode, it can be a limit for timeout: a simple key => value(s) storage is needed to manage multiple settings in PrestaShop.

Reviewer 2: ```suggestion
First, there are multiple situations where a single value has to be stored and be available easily. It can be a boolean to control order mode, it can be a limit for timeout: a simple key => value(s) storage is needed to manage multiple settings in PrestaShop.
```

Secondly, modules might sometimes need a simple storage system, and they should not have to create a new SQL table for such a simple need.

To address these needs, the Configuration storage of PrestaShop uses a SQL table `ps_configuration` and allows simple fetch, update and deletion of values in it.

{{< figure src="/1.7/development/components/img/configuration_sql_structure.png" title="Configuration SQL structure" >}}

> Each value stored in the Configuration component can be controlled through given key, is compatible with Multishop (as it can be overriden for a shop or a group of shop) and has timestamps.

Reviewer 2: ```suggestion
Each value stored in the Configuration component can be controlled through a given key, is compatible with Multishop (as it can be overridden for a shop or a group of shops) and has timestamps.
```

> The `Configuration` class is the interface allowing each module developer to use the component: to get or store data.

Reviewer 3: We should prefer the usage of `PrestaShop\PrestaShop\Àdapter\Configuration` which is integrating the multishop behavior, no? :thinking:

Author: It depends if this documentation aims to explain "what is being used" or "what you should use" I guess

Maybe I can add a comment about this? "You will see the class Configuration used in many places but now please rather use the adapter" ?

Reviewer 3: I think it's a good idea, but this Configuration is not working well with multistore context, we also have to talk about it.

Author: > I think it's a good idea, but this Configuration is not working well with multistore context, we also have to talk about it.

What do you mean? I thought it was one of the piece of multistore that was actually working quite correctly 😅  (compared to other stuff like BO Product page)

Reviewer 3: Oh my bad, the multistore configuration is working well, but afaik, we have to prefer the usage of Adapter instead of native classes. That's the only point :sweat_smile:

Reviewer 2: I agree with @PierreRambaud that we should encourage people to use an Adapter, especially that it's working fine with multi-store now

Reviewer 2: maybe we need to have this as a separate paragraph? to inject adapter through constructor or create an instance of it, because there are no static methods in it

Reviewer: We are supposed to encourage the use of interfaces, not adapters: remember, adapters are meant to go away ;)
...but that would require one interface for each class

Reviewer 2: @eternoendless I was not precise enough, in almost every new code in PrestaShop, where we need, we're injecting @prestashop.adapter.legacy.configuration' 

it has a proper interface and... this is recommended way, right? 🤔

Reviewer: Yep, that's the recommended way, depend on the interface, inject the adapter

## Usage

### Store configuration data

```php

> <?php

Reviewer: No longer necessary
```suggestion
```

Configuration::updateValue(string $key, mixed $value);
```

`$key` will be the reference to your data to reuse later.

> `$value` can be any scalar type (int, string, bool, array...).

Reviewer 2: Most of the time array value means that configuration is multi-lang, we need to point this because if someone has a structure like:

```
0 => 'XYZ',
1 => 'ABC,
```

it will be treated as multi-lang even though the developer might not need it

Reviewer: Also, an array is not a scalar:

> Scalar variables are those containing an int, float, string or bool. Types array, object and resource are not scalar.

https://www.php.net/manual/en/function.is-scalar.php

Reviewer: Follow up from @kpodemski 's comment
```suggestion
`$value` can be any scalar type (int, float, string, bool) or an array.

{{% notice tip %}}
**Multi language**.

By convention, multilanguage values are stored in arrays indexed by language id, like this:

```php
[
    123 => 'Value in some language',
    456 => 'Value in some other language',
]
```
{{% /notice %}}
```

### Check a configuration data set exists

```php

> <?php

Reviewer: ```suggestion
```

Configuration::hasKey(string $key);
```

`$key` is the name of the configuration data to check. The method returns `true`
if the data exists, `false` otherwise.

### Retrieve configuration data

#### Single key

```php

> <?php

Reviewer: ```suggestion
```

Configuration::get(string $key);
```

With `$key` as the data to retrieve.
If the key does not exist, the returned value will be null.

Example:
```php

> <?php

Reviewer: ```suggestion
```

Configuration::get('PS_VERSION_DB');

// returns '1.7.4.0'
```

#### Multiple keys

```php

> <?php

Reviewer: ```suggestion
```

Configuration::getMultiple(array $keys);
```

With `$keys` as an array of keys to retrieve.

This returns an array, containing the values stored in the configuration table or null if a key does not exist.

Example:
```php

> <?php

Reviewer: ```suggestion
```

> Configuration::getMultiple(array('PS_VERSION_DB', 'UNKNOWN_KEY'));

Reviewer 2: ```suggestion
Configuration::getMultiple(['PS_VERSION_DB', 'UNKNOWN_KEY']);
```

I think our CS is setting arrays to short version right?

// returns this array:

> array(
>   'PS_VERSION_DB' => '1.7.4.0',
>   'UNKNOWN_KEY' => null,
> )

Reviewer 2: ```suggestion
[
  'PS_VERSION_DB' => '1.7.4.0',
  'UNKNOWN_KEY' => null,
]
```

```
#### Delete a key

```php

> <?php

Reviewer: ```suggestion
```
